src/components/Tooltip/BaseGenericTooltip/index.native.tsx

```diff
@@ -0,0 +1,132 @@
+import React, {useEffect, useMemo, useRef, useState} from 'react';
+import {Animated, View} from 'react-native';
+// eslint-disable-next-line no-restricted-imports
+import type {Text as RNText, View as RNView} from 'react-native';
+import Text from '@components/Text';
+import useStyleUtils from '@hooks/useStyleUtils';
+import textRef from '@src/types/utils/textRef';
+import viewRef from '@src/types/utils/viewRef';
+import type {BaseGenericTooltipProps} from './types';
+
+// Props will change frequently.
+// On every tooltip hover, we update the position in state which will result in re-rendering.
+// We also update the state on layout changes which will be triggered often.
+// There will be n number of tooltip components in the page.
+// It's good to memoize this one.
+function BaseGenericTooltip({
+    animation,
+    windowWidth,
+    xOffset,
+    yOffset,
+    targetWidth,
+    targetHeight,
+    shiftHorizontal = 0,
+    shiftVertical = 0,
+    text,
+    numberOfLines,
+    maxWidth = 0,
+    renderTooltipContent,
+    shouldForceRenderingBelow = false,
+    shouldForceRenderingLeft = false,
+    wrapperStyle = {},
+}: BaseGenericTooltipProps) {
+    // The width of tooltip's inner content. Has to be undefined in the beginning
+    // as a width of 0 will cause the content to be rendered of a width of 0,
+    // which prevents us from measuring it correctly.
+    const [contentMeasuredWidth, setContentMeasuredWidth] = useState<number>();
+    // The height of tooltip's wrapper.
+    const [wrapperMeasuredHeight, setWrapperMeasuredHeight] = useState<number>();
+    const textContentRef = useRef<RNText>(null);
+    const viewContentRef = useRef<RNView>(null);
+    const rootWrapper = useRef<RNView>(null);
+
+    const StyleUtils = useStyleUtils();
+
+    useEffect(() => {
+        if (!textContentRef.current && !viewContentRef.current) {
+            return;
+        }
+        textContentRef.current?.measure((x, y, width) => setContentMeasuredWidth(width));
+        viewContentRef.current?.measure((x, y, width) => setContentMeasuredWidth(width));
```

Comment:
Actually, how does this logic work? Don't these two lines race with each other? They both set `setContentMeasuredWidth` 🤔

Reply:
Oh, it wasn't clear form the code. I think we can introduce a helper local variable `contentRef = textContentRef.current ?? viewContentRef.current;`. It would be a good place for a comment mentioning that they're mutually exclusive.

The code as is looks like a race condition; let's change this so we don't confuse engineers.